realtek: add support for Panasonic Switch-M8eG PN28080K

musashino205 opened the pull request:
- enable pca953x driver in realtek target
  - to use the LED under the PCA9539PW for system status of OpenWrt while booting
- add support for Panasonic Switch-M8eG PN28080K

musashino205 committed:
realtek: add support for Panasonic Switch-M8eG PN28080K

Panasonic M8eG PN28080K is a 8 + 1 port gigabit switch, based on
RTL8380M.

Specification:

- SoC		: Realtek RTL8380M
- RAM		: DDR3 128 MiB (Winbond W631GG8KB-15)
- Flash		: SPI-NOR 32 MiB (Macronix MX25L25635FMI-10G)
- Ethernet	: 10/100/1000 Mbps x8 + 1
  - port 1-8	: TP, RTL8218B (SoC)
  - port 9	: SFP, RTL8380M (SoC)
- LEDs/Keys	: 7x / 1x
- UART		: RS-232 port on the front panel (connector: RJ-45)
  - 3:TX, 4:GND, 5:GND, 6:RX (pin number: RJ-45)
  - 9600n8
- Power		: 100-240 VAC, 50/60 Hz, 0.5 A
  - Plug	: IEC 60320-C13
- Stock OS	: VxWorks based

Flash instruction using initramfs image:

1.  Prepare the TFTP server with the IP address 192.168.1.111
2.  Rename the OpenWrt initramfs image to "0101A8C0.img" and place it to
    the TFTP directory
3.  Download the official upgrading firmware (ex: pn28080k_v30000.rom)
    and place it to the TFTP directory
4.  Boot M8eG and interrupt the U-Boot with Ctrl + C keys
5.  Execute the following commands and boot with the OpenWrt initramfs
    image

    rtk network on
    tftpboot 0x81000000
    bootm

6.  Backup mtdblock files to the computer by scp or anything and reboot
7.  Interrupt the U-Boot and execute the following commands to re-create
    filesystem in the flash

    ffsmount c:/
    ffsfmt c:/

    this step takes a long time, about ~ 4 mins

8.  Execute the following commands to put the official images to the
    filesystem

    updatert <official image>

    example:

      updatert pn28080k_v30000.rom

    this step takes about ~ 40 secs

9.  Set the environment variables of the U-Boot by the following commands

    setenv loadaddr 0xb4e00000
    setenv bootcmd bootm
    saveenv

10: Download the OpenWrt initramfs image and boot with it

    tftpboot 0x81000000 0101A8C0.img
    bootm

11: On the initramfs image, download the sysupgrade image and perform
    sysupgrade with it

    sysupgrade <imagename>

12: Wait ~ 120 seconds to complete flashing

Note:

- "Switch-M8eG" is a model name, and "PN28080K" is a model number.
  Switch-M8eG has an another (old) model number ("PN28080"), it's not a
  Realtek based hardware.

- Switch-M8eG has a "POWER" LED (Green), but it's not connected to any
  GPIO pin.

- The U-Boot checks the runtime images in the flash when booting and
  fails to execute anything in "bootcmd" variable if the images are not
  exsisting.

- A filesystem is formed in the flash (0x100000-0x1DFFFFF) on the stock
  firmware and it includes the stock images, configuration files and
  checksum files. It's unknown format, can't be managed on the OpenWrt.
  To get the enough space for OpenWrt, move the filesystem to the head
  of "fs_reserved" partition by execution of "ffsfmt" and "updatert".

Back to the stock firmware:

1. Delete "loadaddr" variable and set "bootcmd" to the original value

   on U-Boot:

     setenv loadaddr
     setenv bootcmd 'bootm 0x81000000'

   on OpenWrt:

     fw_setenv loadaddr
     fw_setenv bootcmd 'bootm 0x81000000'

2. Perform reset or reboot

  on U-Boot:

    reset

  on OpenWrt:

    reboot

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
realtek: enable pca953x driver for target

The system status LED on Panasonic Switch-M8eG PN28080K is connected to
a PCA9539PW. To use the LED as a status LED of OpenWrt while booting,
enable the pca953x driver and built-in to the kernel.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

adschm commented on target/linux/realtek/dts/rtl8380_panasonic_m8eg-pn28080k.dts:
```diff
@@ -0,0 +1,280 @@
+		loop_history {
+			label = "green:loop_history";
+			gpios = <&gpio2 11 GPIO_ACTIVE_LOW>;
+		};
```
As always, I'm not really a fan of mixing different "names" for node and label here. If it's full duplex, why not call the led "green:full-duplex" as well so the name will actually explain what it's for?

Why is the label status_eco and the node just "status". And what's any_col? ...

musashino205 replied:
@adschm I used the generic name for node name, and for label, I used the name that is actually printed.

![IMG_20210529_103801](https://user-images.githubusercontent.com/10043798/120054309-208b5280-c06a-11eb-85e7-800d7d45617c.jpg)

> Why is the label status_eco and the node just "status"

It's my bad, I'll change the node name to ```status_eco```.

> what's any_col

If this mode is selected, the port LED where the collision has occurred so far will be turned on.

adschm replied:
Okay, then just using "giga" and "100m" for the node names would be most consistent.

musashino205 replied:
Thank you, I'll update the names.

musashino205 commented:
UPDATE:

- update the node names and dt-labels
  - ```status_(green|amber)``` -> ```status_eco_(green|amber)```
    - dt-label: ```led_status_(green|amber)``` -> ```led_status_eco_(green|amber)```
  - ```collision``` -> ```any_collision```

predators46 commented:
Wao

musashino205 commented:
UPDATE:

- update the node names in LEDs
  - ```gigabit``` -> ```giga```
  - ```100mbps``` -> ```100m```

musashino205 commented:
UPDATE:

- fix node address of ```vlog_data``` partition
  - ```partition@1e20000``` -> ```partition@1e00000```

adschm commented:
Sorry, I won't merge anything non-trivial on realtek.

musashino205 commented:
I see, thank you. 
I sympathize with that attitude towards areas where do not have much knowledge.
However, if all the reactions suddenly stop, I may feel that it is left unattended, so it would be helpful if you could leave some mark  :)

musashino205 commented:
I'll update this PR for 5.10 later.

musashino205 committed:
realtek: enable gpio-restart driver in target

On Panasonic Switch-M8eG PN28080K, a GPIO pin on PCA9539 chip is used
for for hard-reset of the system. To use this, enable gpio-restart
driver and built-in to the kernel.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 commented:
UPDATE:

- fix wrong model name in pca953x commit

musashino205 commented:
I'll re-send the changes to the ML...
